[Docs] Add new_task example; improve: feature store engines descriptions, log_artifact, function handler (#4756)
mlrun/mlrun: docs/runtimes/functions.md

(Functions)=
# Functions 

> All the executions in MLRun are based on serverless functions. Functions are essentially Python objects that know how to run locally or on a Kubernetes cluster. 

Reviewer: ```suggestion
All the executions in MLRun are based on serverless functions. Functions are essentially Python code that can be executed locally or on a Kubernetes cluster. 
```

I don't like the "objects that know how to run..." idiom. The entire idea is that the code doesn't need to know anything about where it runs.

Author: updated

<img src="../_static/images/mlrun-functions.png" alt="mlrun-architecture" width="700"/><br>

MLRun supports numerous real-time and batch runtimes, described in {ref}`function_runtimes`. 
The different function runtimes take care of automatically transforming the code and spec to fully 
managed and elastic services over Kubernetes, which saves significant operational overhead, 
addresses scalability, and reduces infrastructure costs. The function parameters and capabilities 
are explained in more detail in {ref}`create-and-use-functions`.

Function objects are all inclusive, containing the code and all the operational aspects: ({ref}`image <images-usage>`, required packages, 
{ref}`pod resource congifuration (replicas, CPU/GPU/memory limits, volumes, Spot vs. On-demand nodes, pod priority, node affinity) <configuring-job-resources>`, 
[storage](../runtimes/function-storage.html), environment, metadata definitions, etc.). Each function is versioned and stored in the MLRun 
database with a unique hash code, 
and gets a new hash code upon changes.

MLRun functionality such as logging results and artifacts can be added to existing code using the {py:func}`mlrun.handler` decorator, 

> enabled by default. 

Reviewer: Since the handler decorator is now enabled by default (as part of packagers) I would say it's actually a bad practice to use the handler anymore. Instead we should provide details on how to specify these configurations using the `returns` parameter and the input type hints. @guy1992l , @alonmr  - WDYT?

Reviewer 2: Agreed

Author: @theSaarco  - updated text, pls check

- Batch runtimes (based on Kubernetes jobs, {ref}`Spark Operator<spark-operator>`, {ref}`Dask <dask-overview>`, {ref}`Horovod <horovod>`, etc.)

The [different function runtimes](../concepts/functions-overview.html) take care of automatically transforming the code and spec to fully 
managed and elastic services over Kubernetes, which saves significant operational overhead, 
addresses scalability, and reduces infrastructure costs. The function parameters and capabilities are explained in more detail in {ref}`create-and-use-functions`.

Function objects are all inclusive, containing the code and all the operational aspects: ({ref}`image <images-usage>`, required packages, 
{ref}`pod resource congifuration (replicas, CPU/GPU/memory limits, volumes, Spot vs. On-demand nodes, pod priority, node affinity) <configuring-job-resources>`, 
[storage](../runtimes/function-storage.html), environment, metadata definitions, etc.). Each function is versioned and stored in the MLRun database with a unique hash code, 
and gets a new hash code upon changes.

> When you create a function, you specify its name, which appears in the MLRun UI, as well as the Kubernetes pod.

Reviewer: This sentence is not clear - it may be understood from it that you specify its k8s pod, which not only is not correct, it also negates what's written above.
I assume you meant that the the k8s pod used to execute the code appears in the MLRun UI, but why is this relevant here? This sentence deals with what happens when you create a function.

Author: deleted

Depending on the job type, the actual python function 

> can be included in the code as a {py:func}`~mlrun.handler`. You can also use the handler to log results and artifacts, see {ref}`decorators-and-auto-logging`.

Reviewer: This sentence is also confusing. As the `handler` you point to here is the MLRun decorator. I think a better explanation would be something like:

> MLRun functionality such as logging results and artifacts can be added to existing code using the {py:func}`~mlrun.handler` decorator. You can also use MLRun auto-logging to automatically track machine learning processes such as training a model and record results and artifacts seamlessly while they execute. See {ref}`decorators-and-auto-logging`.

Reviewer 2: Worth to note that the handler decorator is now effective by default since the introduction of packagers.

Author: @alonmr  what do you mean "effective by default"? Enabled?

Reviewer 2: https://github.com/mlrun/mlrun/blob/development/mlrun/package/__init__.py#L51

Author: updated to 'enabled by default'

## Function hub

Since function objects are all-inclusive (code, spec, API, and metadata definitions), they can be stored in, 
and reused from, a shared and versioned function hub. This means that different members of the team can produce or 

> consume functions. 

Reviewer: Partially true - it's definitely not easy to share new functions through the hub, and using a custom, private hub, is currently not easily supported. It is easy to consume functions from the hub, maybe we should focus on that...
If we want to emphasize the sharing of functions, maybe it's better to focus on how many users can share the same MLRun project and get access to objects associated with it.

Author: ...that sentence has been there for ages. Good to get it fixed!
